### DISCUSSION TITLE
Connecting `gokrazy` on a zero2W to a Creality Ender3v2 3D printer

### DISCUSSION THREAD

**Discussion by binaek** [gokrazy/gokrazy#285] [Q&A] [answered]
I am trying to connect a RPi Zero2 W to a Creality Ender 3 v2 3D printer.

The idea is to be able to write up small appliance that can queue up and dish out prints without me having to intervene. 

There are a few solutions available that can do this, but it seemed like a good opportunity for me to learn.

[Motherboard Schematic](https://github.com/user-attachments/files/18219132/Creality.4.2.2.-.Schematic.22-9-21.pdf)

From what I can make out, the mother board uses a `CH340` USB<>Serial Converter.

I am trying to connect a RPi 02w USB to the device, but can't seem to figure out how.

`dmesg` provides the following when I connect the printer to my RPi 02W:

```sh
[  433.800984] Indeed it is in host mode hprt0 = 00021501
[  433.988853] usb 1-1: new full-speed USB device number 5 using dwc_otg
[  433.989078] Indeed it is in host mode hprt0 = 00021501
[  434.197902] usb 1-1: New USB device found, idVendor=1a86, idProduct=7523, bcdDevice= 2.64
[  434.197939] usb 1-1: New USB device strings: Mfr=0, Product=2, SerialNumber=0
[  434.197957] usb 1-1: Product: USB Serial
```

It doesn't look like there's any difference in the number of `tty` devices before and after I connect.

However, there's a new file in `/dev/bus/usb/...`

```sh
/tmp/breakglass1809747051 # ls -l /dev/bus/usb/001/*
crw-------    1 0        0         189,   0 Jan  1  1970 /dev/bus/usb/001/001
crw-------    1 0        0         189,   4 Dec 21 21:07 /dev/bus/usb/001/005
```

> The `005` is when I disconnected and connected the printer for the fifth time.

I have hit a bit of a roadblock here.

Not sure where to go from here. Need Help!!!

**Comment by stapelberg** [comment, marked as answer]
Hey @binaek 👋 From your description it sounds like the USB-to-serial Linux driver is missing.

You didn’t say which kernel you are using, but I am suspecting it’s either gokrazy/kernel or gokrazy/kernel.rpi (in the `gok` output, look for the `Kernel directory: ` line if you are unsure).

I have updated both kernels to set the `CONFIG_USB_SERIAL_CH341=y` option, which will hopefully make your device work.

You can update to the newer kernel by running one of these two commands, depending on which kernel you use:

`gok get github.com/gokrazy/kernel@v0.0.0-20241222102143-a155f249fc86`

or:

`gok get github.com/gokrazy/kernel.rpi@v0.0.0-20241222103041-fc92b54edb47`

**Comment by binaek** [comment]
@stapelberg 

The kernel that was getting used was from back in `202408xx` from what I could make out.

Just ran the `gok update` you suggested.

Thanks for the quick turn around

Getting back with results

**Comment by binaek** [comment]
Yep - I see `ttyUSB` now. Thanks @stapelberg ...
